File: src/main/java/com/learnspark/knowledge/repository/KnowledgeEntryRepository.java
package com.learnspark.knowledge.repository;

import com.learnspark.knowledge.entity.KnowledgeEntry;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

/**
 * 知识条目仓库（自动过滤已软删除记录）。
 */
@Repository
public interface KnowledgeEntryRepository extends JpaRepository<KnowledgeEntry, String> {

    /** 按用户查询所有知识条目（按创建时间倒序） */
    List<KnowledgeEntry> findByUserIdOrderByCreatedAtDesc(String userId);

    /** 按来源查询（如某次提交对应的知识条目） */
    Optional<KnowledgeEntry> findBySourceTypeAndSourceId(String sourceType, String sourceId);

    /** 按用户 + 项目查询 */
    List<KnowledgeEntry> findByUserIdAndProjectIdOrderByCreatedAtDesc(String userId, String projectId);

    /** 按用户 + 标题查询（用于 [[双向链接]] 匹配目标条目） */
    Optional<KnowledgeEntry> findByUserIdAndTitle(String userId, String title);

    /** 统计用户知识库条目数（用于 kb_10 徽章与仪表盘统计） */
    long countByUserId(String userId);

    /**
     * 查询内容中包含指定 [[wikilink]] 文本的所有条目（用于新建条目时回扫反链）。
     *
     * <p>利用 @SQLRestriction 自动过滤已删除条目。
     * pattern 应为 {@code [[链接文本]]} 格式。
     */
    @Query("SELECT e FROM KnowledgeEntry e WHERE e.userId = :userId AND " +
           "(e.contentMd LIKE %:pattern% OR e.content LIKE %:pattern%)")
    List<KnowledgeEntry> findEntriesContainingWikilink(@Param("userId") String userId,
                                                       @Param("pattern") String pattern);

    /**
     * MySQL FULLTEXT 全文检索（ngram 分词器，支持中文）。
     *
     * <p>注意：native query 不受 {@code @SQLRestriction} 影响，需手动过滤 deleted_at。
     */
    @Query(value = "SELECT * FROM knowledge_entries " +
            "WHERE user_id = :userId AND deleted_at IS NULL " +
            "AND MATCH(title, content) AGAINST(:q IN NATURAL LANGUAGE MODE)",
            countQuery = "SELECT COUNT(*) FROM knowledge_entries " +
                    "WHERE user_id = :userId AND deleted_at IS NULL " +
                    "AND MATCH(title, content) AGAINST(:q IN NATURAL LANGUAGE MODE)",
            nativeQuery = true)
    Page<KnowledgeEntry> searchByFulltext(@Param("userId") String userId,
                                          @Param("q") String query,
                                          Pageable pageable);
}

File: learnspark-kmp/server/src/main/kotlin/com/learnspark/server/domain/repository/KnowledgeEntryRepository.kt
package com.learnspark.server.domain.repository

import com.learnspark.server.domain.entity.KnowledgeEntry
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.Instant

interface KnowledgeEntryRepository : JpaRepository<KnowledgeEntry, String> {

    fun findByUserIdOrderByUpdatedAtDesc(userId: String, pageable: Pageable): Page<KnowledgeEntry>

    /**
     * 按用户过滤（v1 简化版：不限制 userId 走 DB，依赖 X-User-Id 头层校验）。
     * 阶段 2.4：标签过滤。
     */
    @Query(
        """
        SELECT k FROM KnowledgeEntry k
        WHERE k.userId = :userId
        ORDER BY k.updatedAt DESC
        """
    )
    fun findByUserIdAndOptionalTag(
        @Param("userId") userId: String,
        @Param("tag") tag: String?,
        pageable: Pageable,
    ): Page<KnowledgeEntry>

    /**
     * 阶段 2.4：MySQL FULLTEXT + ngram 全文检索。
     *
     * 注意：FULLTEXT MATCH AGAINST 在 JPQL 中以 nativeQuery 形式表达更直接。
     * 使用 nativeQuery 走 MySQL 原生 MATCH ... AGAINST 语法。
     * booleanMode=true 让 ngram 解析器正常工作（默认 natural language 在中文短词上效果差）。
     */
    @Query(
        value = """
            SELECT * FROM knowledge_entries
            WHERE user_id = :userId
              AND MATCH(title, content) AGAINST (:query IN BOOLEAN MODE)
            ORDER BY updated_at DESC
            LIMIT :#{#pageable.pageSize} OFFSET :#{#pageable.offset}
        """,
        countQuery = """
            SELECT COUNT(*) FROM knowledge_entries
            WHERE user_id = :userId
              AND MATCH(title, content) AGAINST (:query IN BOOLEAN MODE)
        """,
        nativeQuery = true,
    )
    fun fulltextSearch(
        @Param("userId") userId: String,
        @Param("query") query: String,
        pageable: Pageable,
    ): Page<KnowledgeEntry>

    @Modifying
    @Query("UPDATE KnowledgeEntry k SET k.content = :content, k.parseStatus = com.learnspark.server.domain.entity.KnowledgeEntry.ParseStatus.READY, k.parseError = null, k.updatedAt = :now, k.version = k.version + 1 WHERE k.id = :id")
    fun markReady(
        @Param("id") id: String,
        @Param("content") content: String,
        @Param("now") now: Instant,
    ): Int

    @Modifying
    @Query("UPDATE KnowledgeEntry k SET k.parseStatus = com.learnspark.server.domain.entity.KnowledgeEntry.ParseStatus.FAILED, k.parseError = :err WHERE k.id = :id")
    fun markFailed(
        @Param("id") id: String,
        @Param("err") err: String,
    ): Int

    @Modifying
    @Query("UPDATE KnowledgeEntry k SET k.parseStatus = com.learnspark.server.domain.entity.KnowledgeEntry.ParseStatus.PROCESSING WHERE k.id = :id AND k.parseStatus = com.learnspark.server.domain.entity.KnowledgeEntry.ParseStatus.PENDING")
    fun markProcessing(@Param("id") id: String): Int

    /**
     * R3a：按 folderId 列出 knowledge 条目。
     */
    fun findByFolderIdAndUserId(folderId: String, userId: String): List<KnowledgeEntry>
}
